### PR TITLE
KIBA 1354: Implement a "refresh whole collection" flow

### PR DESCRIPTION
**Pull request by Femi-Ogunkola**

### CODE REVIEW

File: api/notd/store/retriever.py [hunk 1]
```diff
@@ -104,6 +104,13 @@ async def get_most_traded_token(self, startDate: datetime.datetime, endDate: dat
         row = result.first()
         return Token(registryAddress=row[TokenTransfersTable.c.registryAddress], tokenId=row[TokenTransfersTable.c.tokenId])
 
+    async def list_tokens_by_collection(self, address: str, connection: Optional[DatabaseConnection] = None) -> List:
```
Comment (krishan711) [inline api/notd/store/retriever.py:107]: this query looks wierd, can you post here the sql it outputs pls

File: api/notd/messages.py [hunk 2]
```diff
@@ -21,6 +21,7 @@ class UpdateTokenMetadataMessageContent(MessageContent):
     _COMMAND = 'UPDATE_TOKEN_METADATA'
     registryAddress: str
     tokenId: str
+    shouldForce: Optional[bool]
```
Comment (krishan711) [inline api/notd/messages.py:24]: do the same for update collection too pls

File: api/notd/token_manager.py [hunk 3]
```diff
@@ -80,7 +80,7 @@ async def update_token_metadatas_deferred(self, collectionTokenIds: List[Tuple[s
             recentlyUpdatedTokenIds = set((tokenMetadata.registryAddress, tokenMetadata.tokenId) for tokenMetadata in recentlyUpdatedTokenMetadatas)
             logging.info(f'Skipping {len(recentlyUpdatedTokenIds)} collectionTokenIds because they have been updated recently.')
             collectionTokenIds = set(collectionTokenIds) - recentlyUpdatedTokenIds
-        messages = [UpdateTokenMetadataMessageContent(registryAddress=registryAddress, tokenId=tokenId).to_message() for (registryAddress, tokenId) in collectionTokenIds]
+        messages = [UpdateTokenMetadataMessageContent(registryAddress=registryAddress, tokenId=tokenId, shouldForce=True).to_message() for (registryAddress, tokenId) in collectionTokenIds]
```
Comment (krishan711) [inline api/notd/token_manager.py:83]: use the param passed in, dont hardcode to True

File: api/notd/token_manager.py [hunk 4]
```diff
@@ -182,3 +182,7 @@ async def update_collection(self, address: str, shouldForce: bool = False) -> No
...
             else:
                 await self.saver.create_collection(connection=connection, address=address, name=retrievedCollection.name, symbol=retrievedCollection.symbol, description=retrievedCollection.description, imageUrl=retrievedCollection.imageUrl, twitterUsername=retrievedCollection.twitterUsername, instagramUsername=retrievedCollection.instagramUsername, wikiUrl=retrievedCollection.wikiUrl, openseaSlug=retrievedCollection.openseaSlug, url=retrievedCollection.url, discordUrl=retrievedCollection.discordUrl, bannerImageUrl=retrievedCollection.bannerImageUrl, doesSupportErc721=retrievedCollection.doesSupportErc721, doesSupportErc1155=retrievedCollection.doesSupportErc1155)
+
+    async def refresh_collection(self, address: str):
+        collectionTokenIds = await self.retriever.list_tokens_by_collection(address=address)
```
Comment (krishan711) [inline api/notd/token_manager.py:187]: you need to update the collection itself as well

File: api/notd/store/retriever.py (hunk 1 above)
Comment (Femi-Ogunkola) [reply api/notd/store/retriever.py:107]: SELECT DISTINCT tbl_token_transfers.registry_address, tbl_token_transfers.token_id 
FROM tbl_token_transfers 
WHERE tbl_token_transfers.registry_address = :registryAddress_1

Comment (Femi-Ogunkola) [reply api/notd/store/retriever.py:107]: SELECT DISTINCT tbl_token_transfers.registry_address, tbl_token_transfers.token_id 
FROM tbl_token_transfers 
WHERE tbl_token_transfers.registry_address = :registryAddress_1
[('0x7509615baf2f9d2eE3274BDa6CC1FC444e0488e8', '2796'), ('0x7509615baf2f9d2eE3274BDa6CC1FC444e0488e8', '3995'), ('0x7509615baf2f9d2eE3274BDa6CC1FC444e0488e8', '3996'), ('0x7509615baf2f9d2eE3274BDa6CC1FC444e0488e8', '3997'), ('0x7509615baf2f9d2eE3274BDa6CC1FC444e0488e8', '5470'), ('0x7509615baf2f9d2eE3274BDa6CC1FC444e0488e8', '5604'), ('0x7509615baf2f9d2eE3274BDa6CC1FC444e0488e8', '6717') ...

File: api/notd/manager.py [hunk 5]
```diff
@@ -155,6 +155,9 @@ async def update_collection_deferred(self, address: str, shouldForce: bool = Fal
     async def update_collection(self, address: str, shouldForce: bool = False) -> None:
         return await self.tokenManager.update_collection(address=address, shouldForce=shouldForce)
 
+    async def refresh_collection(self, address: str, shouldForce: bool = False) -> None:
```
Comment (krishan711) [inline api/notd/manager.py:158]: update_collection_tokens (it should match the message name)

File: api/notd/store/retriever.py (hunk 1 above)
Comment (krishan711) [reply api/notd/store/retriever.py:107]: nice, thanks

File: api/notd/token_manager.py [hunk 6]
```diff
@@ -182,3 +182,8 @@ async def update_collection(self, address: str, shouldForce: bool = False) -> No
...
             else:
                 await self.saver.create_collection(connection=connection, address=address, name=retrievedCollection.name, symbol=retrievedCollection.symbol, description=retrievedCollection.description, imageUrl=retrievedCollection.imageUrl, twitterUsername=retrievedCollection.twitterUsername, instagramUsername=retrievedCollection.instagramUsername, wikiUrl=retrievedCollection.wikiUrl, openseaSlug=retrievedCollection.openseaSlug, url=retrievedCollection.url, discordUrl=retrievedCollection.discordUrl, bannerImageUrl=retrievedCollection.bannerImageUrl, doesSupportErc721=retrievedCollection.doesSupportErc721, doesSupportErc1155=retrievedCollection.doesSupportErc1155)
+
+    async def refresh_collection(self, address: str, shouldForce: bool = False):
```
Comment (krishan711) [inline api/notd/token_manager.py:186]: match function name (comment above)

File: api/notd/notd_message_processor.py [hunk 7]
```diff
@@ -39,11 +40,14 @@ async def process_message(self, message: SqsMessage) -> None:
             return
         if message.command == UpdateTokenMetadataMessageContent.get_command():
             messageContent = UpdateTokenMetadataMessageContent.parse_obj(message.content)
-            await self.notdManager.update_token_metadata(registryAddress=messageContent.registryAddress, tokenId=messageContent.tokenId)
+            await self.notdManager.update_token_metadata(registryAddress=messageContent.registryAddress, tokenId=messageContent.tokenId, shouldForce=messageContent.shouldForce)
             return
         if message.command == UpdateCollectionMessageContent.get_command():
             messageContent = UpdateCollectionMessageContent.parse_obj(message.content)
             await self.notdManager.update_collection(address=messageContent.address)
```
Comment (krishan711) [inline api/notd/notd_message_processor.py:47]: pass in shouldForce here too

File: api/scripts/update_collection_tokens.py [hunk 8]
```diff
@@ -0,0 +1,30 @@
...
+sys.path.insert(0, os.path.dirname(os.path.dirname(__file__)))
+
+import logging
+from typing import Optional
+
+import asyncclick as click
+from core.queues.sqs_message_queue import SqsMessageQueue
+
+sys.path.insert(0, os.path.dirname(os.path.dirname(__file__)))
+from api.notd.messages import UpdateCollectionTokensMessageContent
```
Comment (krishan711) [inline api/scripts/update_collection_tokens.py:14]: this didnt work

**Review by krishan711 — APPROVED** [review]

**Pull request merged**